### PR TITLE
 [780] Add new tool to create a full FeatureTyping from PartUsage 

### PR DESCRIPTION
**Pull request by jerome-obeo**
# PLEASE READ ALL ITEMS AND CHECK ONLY RELEVANT CHECKBOXES BELOW

## Project management

- [x] Has the pull request been added to the relevant milestone?
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `type:`)
- [ ] Have the relevant issues been added to the same project milestone as the pull request?

## Changelog and release notes

- [x] Has the `CHANGELOG.adoc` + `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc` + `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?
- [ ] In case of a change with a visual impact, are there any screenshots in the `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?
- [ ] In case of a key change, has the change been added to `Key highlights` section in `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc` + `doc/content/modules/user-manual/pages/release-notes/YYYY.MM.0.adoc`?

## Documentation

- [ ] Have you included an update of the [documentation](https://doc.mbse-syson.org) in your pull request? Please ask yourself if an update (installation manual, user manual, developer manual...) is needed and add one accordingly.

## Tests

- [x] Is the code properly tested? Any pull request (fix, enhancement or new feature) should come with a test (or several). It could be unit tests, integration tests or cypress tests depending on the context. Only doc and releng pull request do not need for tests.

### CODE REVIEW

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    private final DiagramBuilders diagramBuilderHelper = new DiagramBuilders();
+
+    private final ViewBuilders viewBuilderHelper = new ViewBuilders();
+
+    private final PartUsage partUsage;
+
+    private final IDescriptionNameGenerator nameGenerator;
+
+    public PartUsageFeatureTypingNodeToolProvider(PartUsage partUsage, IDescriptionNameGenerator nameGenerator) {
+        this.partUsage = partUsage;
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:49]: I think this parameter should also has the guard `Objects.requireNonNull()`.
Do you have a reason to not added it?

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ */
+public class PartUsageFeatureTypingNodeToolProvider implements INodeToolProvider {
+
+    private final DiagramBuilders diagramBuilderHelper = new DiagramBuilders();
+
+    private final ViewBuilders viewBuilderHelper = new ViewBuilders();
+
+    private final PartUsage partUsage;
+
+    private final IDescriptionNameGenerator nameGenerator;
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:46]: I would prefer `descriptionNameGenerator`

File: backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java [hunk 3]
```diff
@@ -761,4 +767,38 @@ private boolean isPart(Element element) {
     private boolean isAction(Element element) {
         return element instanceof ActionUsage || element instanceof ActionDefinition;
     }
+
+    /**
+     * Return an instance of the actual {@link PartDefinition} associated to the given {@link PartUsage}.
+     *
```
Comment (AxelRICHARD) [inline backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java:773]: Please explain here that the PartUsage parameter could be one of its substype.
That's the whole purpose of this method.

File: backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java [hunk 4]
```diff
@@ -761,4 +767,38 @@ private boolean isPart(Element element) {
...
+     *         an instance of PartUsage
+     * @return an actual instance of the part definition associated to the given part usage.
+     */
+    public Definition createRightPartDefinitionFrom(PartUsage partUsage) {
+        return (Definition) SysmlFactory.eINSTANCE.create(this.getRightPartDefinitionEClassFrom(partUsage));
+    }
+
+    /**
+     * Return the PartDefinition {@link EClass} corresponding to the given PartUsage.
+     * @param partUsage a PartUsage instance
```
Comment (AxelRICHARD) [inline backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java:784]: Please explain here that the PartUsage parameter could be one of its substype.
That's the whole purpose of this method.

File: backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java [hunk 5]
```diff
@@ -761,4 +767,38 @@ private boolean isPart(Element element) {
...
     }
+
+    /**
+     * Return an instance of the actual {@link PartDefinition} associated to the given {@link PartUsage}.
+     *
+     * @param partUsage
+     *         an instance of PartUsage
+     * @return an actual instance of the part definition associated to the given part usage.
+     */
+    public Definition createRightPartDefinitionFrom(PartUsage partUsage) {
```
Comment (AxelRICHARD) [inline backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java:778]: createRightPartDefinitionFrom -> createPartDefinitionFrom

File: backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java [hunk 6]
```diff
@@ -761,4 +767,38 @@ private boolean isPart(Element element) {
...
+        return (Definition) SysmlFactory.eINSTANCE.create(this.getRightPartDefinitionEClassFrom(partUsage));
+    }
+
+    /**
+     * Return the PartDefinition {@link EClass} corresponding to the given PartUsage.
+     * @param partUsage a PartUsage instance
+     * @return the EClass of the PartDefinition associated to the given PartUsage.
+     */
+    public EClass getRightPartDefinitionEClassFrom(PartUsage partUsage) {
```
Comment (AxelRICHARD) [inline backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java:787]: getRightPartDefinitionEClassFrom -> getPartDefinitionEClassFrom

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    public NodeTool create(IViewDiagramElementFinder cache) {
+        var builder = this.diagramBuilderHelper.newNodeTool();
+
+        var creationFeatureTypingServiceCall = this.viewBuilderHelper.newChangeContext()
+                .expression(AQLUtils.getSelfServiceCallExpression("createPartDefinitionAndFeatureTyping"))
+                .build();
+
+        NodeDescription nodeDescription = cache.getNodeDescription(this.getNodeName()).orElse(null);
+
+        var createView = this.diagramBuilderHelper.newCreateView()
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:63]: What happens if nodeDescription is null?

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                .build();
+
+        return builder.name(this.nameGenerator.getCreationToolName(SysmlPackage.eINSTANCE.getFeatureTyping()))
+                .iconURLsExpression("/icons/full/obj16/FeatureTyping.svg")
+                .body(rootChangContext)
+                .build();
+    }
+
+    private String getNodeName() {
+        var utilService = new UtilService();
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:82]: I would prefer to have the `utilService = new UtilService();` declared as a class field.

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                .children(creationFeatureTypingServiceCall, createView.build())
+                .build();
+
+        return builder.name(this.nameGenerator.getCreationToolName(SysmlPackage.eINSTANCE.getFeatureTyping()))
+                .iconURLsExpression("/icons/full/obj16/FeatureTyping.svg")
+                .body(rootChangContext)
+                .build();
+    }
+
+    private String getNodeName() {
```
Comment (AxelRICHARD) [inline backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:81]: I'm not sure this method is still relevant with only one line.

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java (hunk 1 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:49]: Done

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java (hunk 2 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:46]: Done

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java (hunk 7 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:63]: Changed.
The tool is not created if we are unable to find the definition node description associated with a usage.

File: backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java (hunk 9 above)
Comment (jerome-obeo) [reply backend/views/syson-diagram-common-view/src/main/java/org/eclipse/syson/diagram/common/view/tools/PartUsageFeatureTypingNodeToolProvider.java:81]: Due to previous changes, this no longer exists.

File: backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java (hunk 3 above)
Comment (jerome-obeo) [reply backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java:775]: Done

File: backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java (hunk 4 above)
Comment (jerome-obeo) [reply backend/services/syson-services/src/main/java/org/eclipse/syson/services/UtilService.java:789]: Done

**Pull request merged**